Add support for HOCON config

davidfrickert opened the pull request:
Fixes #107

**TODO: add integration tests**

radcortez commented on hocon/deployment/src/main/java/io/quarkus/config/hocon/deployment/ConfigHoconProcessor.java:
```diff
@@ -0,0 +1,67 @@
+    }
+
+    @BuildStep
+    public void bootstrap(
+            BuildProducer<AdditionalBootstrapConfigSourceProviderBuildItem> additionalBootstrapConfigSourceProvider,
+            BuildProducer<StaticInitConfigSourceProviderBuildItem> staticInitConfigSourceProvider) {
+        additionalBootstrapConfigSourceProvider.produce(new AdditionalBootstrapConfigSourceProviderBuildItem(
+                ApplicationHoconConfigSourceLoader.InFileSystem.class.getName()));
+        additionalBootstrapConfigSourceProvider.produce(new AdditionalBootstrapConfigSourceProviderBuildItem(
+                ApplicationHoconConfigSourceLoader.InClassPath.class.getName()));
+        staticInitConfigSourceProvider.produce(new StaticInitConfigSourceProviderBuildItem(
+                ApplicationHoconConfigSourceLoader.InFileSystem.class.getName()));
+        staticInitConfigSourceProvider.produce(new StaticInitConfigSourceProviderBuildItem(
+                ApplicationHoconConfigSourceLoader.InClassPath.class.getName()));
```
I know we have this in the YAML source, but I recently submitted a PR to use our builder step. I plan to deprecate all of these. Check here:
https://github.com/quarkusio/quarkus/pull/30356/files#diff-6d620b5b84d3f7ea9c637cc33bcc75219b91df94f4d4ec872b410837607c451f
https://github.com/quarkusio/quarkus/pull/30356/files#diff-0f949a92f29bc67d8932bf3f9b014337db84d24b3e5bd774ff5926602689ae4f

davidfrickert commented:
Used the builder instead of bootstrap and added integration tests.

davidfrickert commented:
Should I also add a document in the docs module?

davidfrickert commented:
Can you check my comments? @radcortez
To see if we can get this merged or if something can be improved.

radcortez commented on hocon/deployment/pom.xml:
```diff
@@ -0,0 +1,66 @@
+  <artifactId>quarkus-config-hocon-deployment</artifactId>
+
+  <properties>
+    <maven.compiler.source>17</maven.compiler.source>
+    <maven.compiler.target>17</maven.compiler.target>
```
Do we need to fix this to be only 17?

davidfrickert replied:
not really, probably something added by IDE, will remove

radcortez commented on hocon/deployment/pom.xml:
```diff
@@ -0,0 +1,66 @@
+    <dependency>
+      <groupId>io.quarkiverse.config</groupId>
+      <artifactId>quarkus-config-hocon</artifactId>
+      <version>1.1.1-SNAPSHOT</version>
```
To reference the runtime dependency without fixing the version, use `${project.version}`

radcortez commented on hocon/deployment/src/test/java/io/quarkus/config/yaml/deployment/ApplicationHoconHotDeploymentTest.java:
```diff
@@ -0,0 +1,41 @@
+package io.quarkus.config.yaml.deployment;
```
Package is using `yaml`

radcortez commented on hocon/runtime/pom.xml:
```diff
@@ -0,0 +1,78 @@
+  <artifactId>quarkus-config-hocon</artifactId>
+
+  <properties>
+    <maven.compiler.source>17</maven.compiler.source>
+    <maven.compiler.target>17</maven.compiler.target>
```
Do we need to fix this to be only 17?

davidfrickert replied:
also removed this one, there is no need for 17 yes

radcortez commented on hocon/runtime/src/main/resources/META-INF/services/org.eclipse.microprofile.config.spi.ConfigSourceProvider:
```diff
@@ -0,0 +1,2 @@
+io.quarkus.config.hocon.runtime.ApplicationHoconConfigSourceLoader$InFileSystem
+io.quarkus.config.hocon.runtime.ApplicationHoconConfigSourceLoader$InClassPath
```
With the builder approach, the ServiceLoader registration is no longer required since we register the sources directly with the builder.

radcortez commented on hocon/runtime/src/test/java/io/quarkus/config/hocon/runtime/ApplicationHoconTest.java:
```diff
@@ -0,0 +1,54 @@
+import io.smallrye.config.SmallRyeConfigBuilder;
+
+/**
+ * Test the YAML config provider (plan JUnit). We aren't re-testing the whole config source
```
YAML

radcortez commented on pom.xml:
```diff
@@ -26,6 +27,7 @@
         <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
         <project.reporting.outputEncoding>UTF-8</project.reporting.outputEncoding>
         <quarkus.version>2.15.3.Final</quarkus.version>
+        <smallrye-config.version>3.1.1</smallrye-config.version>
```
This has to align with the SmallRye Config version coming from Quarkus.

davidfrickert replied:
It seems that the version coming from Quarkus is 2.13.1 so I changed it to that one

radcortez commented on integration-tests/hocon/src/main/resources/application.conf:
```diff
@@ -0,0 +1,7 @@
+quarkus {
+    application.name = consul-test
```
Consul

radcortez commented:
> Can you check my comments? @radcortez
> To see if we can get this merged or if something can be improved.

Thanks for the PR. Highly appreciated it :)

Left a few comments for you to consider. Thanks!

davidfrickert commented:
Thanks for the comments. Will fix soon the copy pastes, my bad.

davidfrickert commented:
I think I've fixed everything @radcortez 
Not sure if there's a way to fetchthe `smallrye-config` version directly from Quarkus instead of having it here, so I left that one unresolved if you have any ideas.

In another note, as there might be a potential dependency convergence problems with Quarkus, shouldn't this project use maven enforcer plugin?

radcortez commented:
Yes, the version property is defined in the Quarkus BOM, but I think you cannot access properties from such artifacts, just the dependencies.

Feel free to add the enforcer plugin. Thanks!

davidfrickert commented:
It seems that the  `ApplicationHoconTest` was failing after removing the ConfigSourceProvider declarations in `META-INF`.
Fixed the test to use the `HoconConfigBuilder` and it is passing now.
Also fixed the modules order (hocon module has to be before integration-tests).
And added enforcer plugin, although, for this specific case it will be useless since apparently dependencies coming from bom override any local dependencies (i.e., there is no dependency convergence to detect), but it's a nice to have in the project.

Should be good to go @radcortez !

radcortez commented on integration-tests/hocon/src/test/java/io/quarkus/it/consul/config/GreetingResourceTest.java:
```diff
@@ -0,0 +1,26 @@
+package io.quarkus.it.consul.config;
```
Consul package.

radcortez commented:
Found another copy-paste package name :)

Also, CI is not happy. It failed the native test.

davidfrickert commented:
Fixed! 
Hmm I see, I didn't test with native. Any idea what might be happening?

radcortez commented:
No idea... I would have to look.

davidfrickert commented:
I ran `mvn verify -Dnative` and it indeed fails on the native integration test.
Just to be sure, (locally) I reverted from the builder approach to the older bootstrap approach and the native integration test succeeded.

Could you have a look to see why the builder is having problems? I've no clue how to debug native image problems.

If not, maybe we can merge with the bootstrap approach instead of builder?

radcortez commented:
In this case, we do need to either register the hocon files for the native image, or register the source for build time so that the values can be recorded and we can drop the file.

I've pushed a change to register the sources for build time, which should fix the issue.

radcortez commented:
Merged. Thank you very much for all the work :)

davidfrickert commented:
No worries. Thanks for the review and help.
